File: clients/go/model/player_message_game.go
package model

import "io"
import . "aicup2019/stream"

type PlayerMessageGame interface {
    Write(writer io.Writer)
}
func ReadPlayerMessageGame(reader io.Reader) PlayerMessageGame {
    switch ReadInt32(reader) {
        case 0:
            return ReadPlayerMessageGameCustomDataMessage(reader)
        case 1:
            return ReadPlayerMessageGameActionMessage(reader)
    }
    panic("Unexpected discriminant value")
}

type PlayerMessageGameCustomDataMessage struct {
    Data CustomData
}
func NewPlayerMessageGameCustomDataMessage(data CustomData) PlayerMessageGameCustomDataMessage {
    return PlayerMessageGameCustomDataMessage {
        Data: data,
    }
}
func ReadPlayerMessageGameCustomDataMessage(reader io.Reader) PlayerMessageGameCustomDataMessage {
    result := PlayerMessageGameCustomDataMessage {}
    result.Data = ReadCustomData(reader)
    return result
}
func (value PlayerMessageGameCustomDataMessage) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    value.Data.Write(writer)
}

type PlayerMessageGameActionMessage struct {
    Action Versioned
}
func NewPlayerMessageGameActionMessage(action Versioned) PlayerMessageGameActionMessage {
    return PlayerMessageGameActionMessage {
        Action: action,
    }
}
func ReadPlayerMessageGameActionMessage(reader io.Reader) PlayerMessageGameActionMessage {
    result := PlayerMessageGameActionMessage {}
    result.Action = ReadVersioned(reader)
    return result
}
func (value PlayerMessageGameActionMessage) Write(writer io.Writer) {
    WriteInt32(writer, 1)
    value.Action.Write(writer)
}

File: clients/cpp/model/PlayerMessageGame.cpp
#include "PlayerMessageGame.hpp"


PlayerMessageGame::CustomDataMessage::CustomDataMessage() { }
PlayerMessageGame::CustomDataMessage::CustomDataMessage(std::shared_ptr<CustomData> data) : data(data) { }
PlayerMessageGame::CustomDataMessage PlayerMessageGame::CustomDataMessage::readFrom(InputStream& stream) {
    PlayerMessageGame::CustomDataMessage result;
    result.data = CustomData::readFrom(stream);
    return result;
}
void PlayerMessageGame::CustomDataMessage::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    data->writeTo(stream);
}
std::string PlayerMessageGame::CustomDataMessage::toString() const {
    return std::string("PlayerMessageGame::CustomDataMessage") + "(" +
        data->toString() +
        ")";
}

PlayerMessageGame::ActionMessage::ActionMessage() { }
PlayerMessageGame::ActionMessage::ActionMessage(Versioned action) : action(action) { }
PlayerMessageGame::ActionMessage PlayerMessageGame::ActionMessage::readFrom(InputStream& stream) {
    PlayerMessageGame::ActionMessage result;
    result.action = Versioned::readFrom(stream);
    return result;
}
void PlayerMessageGame::ActionMessage::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    action.writeTo(stream);
}
std::string PlayerMessageGame::ActionMessage::toString() const {
    return std::string("PlayerMessageGame::ActionMessage") + "(" +
        action.toString() +
        ")";
}
std::shared_ptr<PlayerMessageGame> PlayerMessageGame::readFrom(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return std::shared_ptr<PlayerMessageGame::CustomDataMessage>(new PlayerMessageGame::CustomDataMessage(PlayerMessageGame::CustomDataMessage::readFrom(stream)));
    case 1:
        return std::shared_ptr<PlayerMessageGame::ActionMessage>(new PlayerMessageGame::ActionMessage(PlayerMessageGame::ActionMessage::readFrom(stream)));
    default:
        throw std::runtime_error("Unexpected discriminant value");
    }
};
